src/appleseed/foundation/image/oiioimagefilewriter.cpp

```diff
@@ -0,0 +1,515 @@
+    }
+}
+
+void OIIOImageFileWriter::set_image_output_format(const PixelFormat output_pixel_format)
+{
+    if (get_image_count() == 0)
+        throw ExceptionIOError("No images available!");
+
+    m_images->m_spec.back().set_format(convert_pixel_format(output_pixel_format));
+}
+
+void OIIOImageFileWriter::set_image_channels(
+    const size_t    channel_count,
+    const char**    channel_names)
+{
+    if (get_image_count() == 0)
+        throw ExceptionIOError("No images available!");
+    if (channel_count < 0)
+        throw ExceptionIOError("channel_count must be superior to 0");
+    if (channel_names == nullptr)
+        throw ExceptionIOError("Valid channel_names must be provide!");
+
+    OIIO::ImageSpec& spec = m_images->m_spec.back();
+
+    spec.nchannels = static_cast<int>(channel_count);
+
+    for (size_t i = 0; i < channel_count; i++)
+    {
+        const char* name = channel_names[i];
+
+        spec.channelnames.push_back(name);
+
+        if (strcmp(name, "A"))
+            spec.alpha_channel = static_cast<int>(i);
+    }
+}
+
+void OIIOImageFileWriter::set_image_spec(void)
+{
+    if (get_image_count() == 0)
+        throw ExceptionIOError("No images available!");
+
+    const CanvasProperties& props = m_images->m_canvas.back()->properties();
+    OIIO::ImageSpec& spec = m_images->m_spec.back();
+
+    // Size of the data of the image.
+    spec.width = static_cast<int>(props.m_canvas_width);
+    spec.height = static_cast<int>(props.m_canvas_height);
+
+    // Origin of the pixel data of the image.
+    spec.x = 0;
+    spec.y = 0;
+
+    // Full size of the data of the image.
+    spec.full_width = spec.width;
+    spec.full_height = spec.height;
+
+    // Origin of the pixel data of the full image.
+    spec.full_x = spec.x;
+    spec.full_y = spec.y;
+
+    // Size of a tile.
+    if (m_writer->supports("tiles"))
+    {
+        spec.tile_width = static_cast<int>(props.m_tile_width);
+        spec.tile_height = static_cast<int>(props.m_tile_height);
+    }
+    else
+    {
+        spec.tile_width = 0;
+        spec.tile_height = 0;
+    }
+
+    // Number of channels.
+    const char* channel_names[] = { "R", "G", "B", "A" };
+    set_image_channels(props.m_channel_count, channel_names);
+
+    // Format of the pixel data.
+    set_image_output_format(props.m_pixel_format);
+}
+
+void OIIOImageFileWriter::set_exr_image_attributes(const ImageAttributes& image_attributes)
+{
+    OIIO::ImageSpec spec = m_images->m_spec.back();
+
+    if (image_attributes.exist("dwa_compression_lvl"))
+        spec.attribute("openexr:dwaCompressionLevel", image_attributes.get<float>("dwa_compression_lvl"));
+
+    if (image_attributes.exist("white_xy_chromaticity") &&
+        image_attributes.exist("red_xy_chromaticity") &&
+        image_attributes.exist("green_xy_chromaticity") &&
+        image_attributes.exist("blue_xy_chromaticity"))
+    {
+        float chromaticities[8];
+        memset(chromaticities, 0, 8 * sizeof(float));
```

Comment:
You need to include `<cstring>` for this.

Reply:
done